src/odemis/driver/piezomotor.py

```diff
@@ -0,0 +1,453 @@
+# -*- coding: utf-8 -*-
+'''
+Created on 18 Mar 2020
+
+@author: Philip Winkler
+
+Copyright © 2020, Philip Winkler, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms
+of the GNU General Public License version 2 as published by the Free Software
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+'''
+from __future__ import division
+
+import glob
+import logging
+from odemis import model
+from odemis.model import Actuator, CancellableThreadPoolExecutor
+from odemis.util import to_str_escape
+import os
+import threading
+import time
+import serial
+from threading import Thread
+import copy
+from odemis.model import Axis, HwComponent
+
+BAUDRATE = 115200
+WFM_STEPSIZE = 5e-6  # m / step
+USTEPS_PER_WFM = 8192
+ENCODER_STEPSIZE = 1e-9  # m / step
+
+EOL = b'\r'  # 0x0D, carriage return
+
+# Waveforms
+WAVEFORM_RHOMB = 1  # fast max speed
+WAVEFORM_DELTA = 2  # preferred, higher accuracy
+WAVEFORM_PARK = 4  # power off
+
+
+class PMAxis(Axis, HwComponent):
+    def __init__(self, name, role, canAbs=True, choices=None, unit=None,
+                 range=None, speed=None, canUpdate=False, *args, **kwargs):
+        HwComponent.__init__(self, name, role, *args, **kwargs)
+        Axis.__init__(self, canAbs=canAbs, choices=choices, unit=unit,
+                      range=range, speed=speed, canUpdate=canUpdate)
+
+
+class PMD401(Actuator):
+    """
+    This represents the PMD401 motor controller for the PiezoMotor LEGS motors.
+    The specification for the hardware interface can be found in the document
+     "PiezoMotor_PMD401_Technical_Manual.pdf".
+
+    The class provides only part of the functionality of the model.Actuator interface. It fully
+    implements the functions moveRel and stop. However, there is no reference function and
```

Comment:
I didn't really understand the section on index mode in the manual.
> The I command works the same (as J) but will only run if index mode is activated and index has not been detected.

I would have thought that index mode allows you to do an absolute move (go to position x), but it seems to work the same J, so with relative moves. I have no clue what "index has not been detected" means.

It seems to be possible to get the expected position with the read command `N?`, so then for an absolute move we can just subtract this from the target and send the relative move command. However, it is still not clear to me how to implement referencing. Just move in one direction until we get to the limit switch and then reset the index?

Referencing is not needed for closed-loop mode, right?

Reply:
Are we ever planning to use open-loop mode anyway? Since we have encoders for all axes, I thought it wasn't really necessary.

Reply:
Reading the chapter 4.5 of the documentation, I understand that this is a quite low-level referencing procedure. The `Nx` command set the device for reference search, and the `I` command actually tells the device to search. You have to build your referencing procedure based on this commands. Depending on the hardware the referencing procedure might defer. In particular, the physical position of the index matters. An "easy" hardware is were the index is at the end of the axis (let's say at the lowest position). In this case, you could do something like:
1. start the index mode for search of index
2. jog in reverse direction (because you know that the index is "before" the current position
3. wait until the index is found, or stop after too much time has passed.

That would translate into something like:
`N4`, `I-16`, `N`..... `S`

Note that this is a really basic referencing procedure. You typically want to do a fast search, and then do a second search at slow speed, to locate the index more precisely. Also, if the index is somewhere in the middle, you'll need to handle the fact you might have picked the wrong direction (in which case, after while you'll bump into the limit switch) and in such case, continue by going in the other direction.

After the referencing procedure, typically what we want is that the index position is considered position 0, for the absolute moves. Also, the axis will be physically at this position (but it doesn't really have to be). 

The driver could provide different type of referencing procedure, defined as a parameter per axis (as we do in the TMCM). That said, hopefully for the hardware we have we can use the same procedure for all the axes... so only one procedure would need to be implemented for now.
I'd suggest to start roughly implementing the simple procedure that I've described, and once the hardware is sufficiently define, adjust the code.